include/mfmg/dealii_operator.hpp

```diff
@@ -28,413 +31,161 @@
+class SmootherOperator : public Operator<VectorType>
 {
 public:
   using vector_type = VectorType;
   using matrix_type = dealii::TrilinosWrappers::SparseMatrix;
+  using operator_type = Operator<vector_type>;
 
-private:
-  enum SmootherType
-  {
-    SGS,
-    GS,
-    JACOBI,
-    ILU
-  };
-
-public:
-  DealIISmootherOperator(const matrix_type &matrix,
-                         std::shared_ptr<boost::property_tree::ptree> params)
-      : _matrix(matrix)
-  {
-    std::string prec_name = params->get<std::string>("preconditioner: type",
-                                                     "Symmetric Gauss-Seidel");
-    _prec_type = string2type(prec_name);
-    initialize(_prec_type);
-  }
+  SmootherOperator(matrix_type const &matrix,
+                   std::shared_ptr<boost::property_tree::ptree> params);
 
   virtual size_t m() const override final { return _matrix.m(); }
 
   virtual size_t n() const override final { return _matrix.n(); }
 
-  void apply(const vector_type &b, vector_type &x) const
-  {
-    // r = -(b - Ax)
-    vector_type r(b);
-    _matrix.vmult(r, x);
-    r.add(-1., b);
-
-    // x = x + B^{-1} (-r)
-    vector_type tmp(x);
-    smoother_vmult(tmp, r);
-    x.add(-1., tmp);
-  }
+  virtual void apply(vector_type const &b, vector_type &x) const override final;
 
-private:
-  SmootherType string2type(const std::string &prec_name) const
-  {
-    // Make parameters case-insensitive
-    std::string prec_name_lower = prec_name;
-    std::transform(prec_name_lower.begin(), prec_name_lower.end(),
-                   prec_name_lower.begin(), ::tolower);
-    if (prec_name_lower == "symmetric gauss-seidel")
-      return SGS;
-    else if (prec_name_lower == "gauss-seidel")
-      return GS;
-    else if (prec_name_lower == "jacobi")
-      return JACOBI;
-    else if (prec_name_lower == "ilu")
-      return ILU;
-    else
-      throw std::runtime_error("Unknown smoother name: \"" + prec_name_lower +
-                               "\"");
-  }
-  void initialize(const SmootherType prec_type)
-  {
-    switch (prec_type)
-    {
-    case SGS:
-      _sgs_smoother.initialize(_matrix);
-      break;
-    case GS:
-      _gs_smoother.initialize(_matrix);
-      break;
-    case JACOBI:
-      _jacobi_smoother.initialize(_matrix);
-      break;
-    case ILU:
-      _ilu_smoother.initialize(_matrix);
-      break;
-    };
-  }
+  virtual std::shared_ptr<operator_type> transpose() const override final;
 
-  void smoother_vmult(vector_type &x, const vector_type &b) const
-  {
-    {
-      switch (_prec_type)
-      {
-      case SGS:
-        _sgs_smoother.vmult(x, b);
-        break;
-      case GS:
-        _gs_smoother.vmult(x, b);
-        break;
-      case JACOBI:
-        _jacobi_smoother.vmult(x, b);
-        break;
-      case ILU:
-        _ilu_smoother.vmult(x, b);
-        break;
-      };
-    }
-  }
+  virtual std::shared_ptr<operator_type>
+  multiply(operator_type const &operator_b) const override final;
```

Comment:
These functions really don't belong to smoother, and should be gone once `MatrixOperator` is moved out of `Operator` concept.